### PR TITLE
Bump Ruby to 3.2.2

### PR DESCRIPTION
**Pull request by tvararu**
### Context

- Ticket: [CPDLP-1913, kinda](https://dfedigital.atlassian.net/browse/CPDLP-1913)

### Changes proposed in this pull request

Update the application to Ruby 3.2.0.

Ruby 3.2.0 brings [substantial performance improvements](https://www.ruby-lang.org/en/news/2022/12/25/ruby-3-2-0-released/).

### Guidance to review

WIP until I get the tests to pass fully.

### CODE REVIEW

**Comment by tvararu** [comment]
Couldn't get this finished in time, but, here's where I got.

Keyword arguments seem to get a slightly different treatment in Ruby 3.2.0.

Let's look at a failure on `bin/bundle exec rspec spec/requests/admin/suppliers/supplier_users_spec.rb`:

```ruby
  7) Admin::SupplierUsers POST /admin/suppliers/new Sets the correct lead provider
     Failure/Error:
       def welcome_email(user:, lead_provider_name:, start_url:)
         template_mail(
           WELCOME_EMAIL_TEMPLATE,
           to: user.email,
           rails_mailer: mailer_name,
           rails_mail_template: action_name,
           personalisation: {
             name: user.full_name,
             lead_provider_name:,
             start_url:,

     ArgumentError:
       wrong number of arguments (given 1, expected 0; required keywords: user, lead_provider_name, start_url)
     # ./app/mailers/lead_provider_mailer.rb:8:in `welcome_email'
     # ./app/models/lead_provider_profile.rb:14:in `block in create_lead_provider_user'
     # ./app/models/lead_provider_profile.rb:10:in `create_lead_provider_user'
     # ./app/forms/supplier_user_form.rb:26:in `save!'
     # ./app/controllers/admin/suppliers/supplier_users_controller.rb:67:in `create'
     # ./app/middlewares/time_traveler.rb:9:in `call'
     # ./spec/requests/admin/suppliers/supplier_users_spec.rb:185:in `when_i_confirm_my_choices'
     # ./spec/requests/admin/suppliers/supplier_users_spec.rb:43:in `block (3 levels) in <top (required)>'

Finished in 3 minutes 57.9 seconds (files took 2.41 seconds to load)
16 examples, 7 failures

Failed examples:

...etc...
rspec ./spec/requests/admin/suppliers/supplier_users_spec.rb:38 # Admin::SupplierUsers POST /admin/suppliers/new Sets the correct lead provider
```

`LeadProviderProfile` calls `welcome_email`:

```ruby
  def self.create_lead_provider_user(full_name, email, lead_provider, start_url)
    ActiveRecord::Base.transaction do
      user = User.create!(full_name:, email:)
      LeadProviderProfile.create!(user:, lead_provider:)
-->   LeadProviderMailer.welcome_email(user:, lead_provider_name: lead_provider.name, start_url:).deliver_now
    end
  end
```

Which is defined in `LeadProviderMailer` as:

```ruby
  def welcome_email(user:, lead_provider_name:, start_url:)
```

I can't quite figure out what's wrong with the callsite/way of defining the arguments list in the method definition.

I tried `LeadProviderMailer.new.welcome_email(user:, etc)` which fixes the arguments issue, but can't be `.deliver`ed.

This looks interesting:

```ruby
pry(LeadProviderProfile)> LeadProviderMailer.new.welcome_email({ :user => user }).deliver
ArgumentError: wrong number of arguments (given 1, expected 0; required keywords: user,
lead_provider_name, start_url)
pry(LeadProviderProfile)> LeadProviderMailer.new.welcome_email(user:).deliver
ArgumentError: missing keywords: :lead_provider_name, :start_url
```

The first is failing with an argument error that says that `user` isn't present. The second isn't. 

The same is not true when we `.new` it; the two syntaxes fail in the same way:

```ruby
[1] pry(LeadProviderProfile)> LeadProviderMailer.welcome_email(user:, lead_provider_name: lead_provider.name, start_url:).deliver_now
ArgumentError: wrong number of arguments (given 1, expected 0; required keywords: user,
lead_provider_name, start_url)
[2] pry(LeadProviderProfile)> LeadProviderMailer.welcome_email({ :user => user, :lead_provider_name => lead_provider.name, :start_url => start_url }).deliver_now
ArgumentError: wrong number of arguments (given 1, expected 0; required keywords: user,
lead_provider_name, start_url)
```

Maybe `mail-notify` is doing something with the arguments being passed into the class methods that inherit from `Mail::Notify::Mailer` that Ruby 3.2.0 doesn't like. If it is, I'm curious why [TVS didn't have a similar issue](https://github.com/DFE-Digital/teaching-vacancies/pull/5716).

**Comment by peteryates** [comment]
This is an interesting one. So locally I merged this into my Rails 7 branch (that's been waiting in the wings for several months 😂) and the tests pass.

Not sure what the actual issue is but it's going to be a fun one. Here's what `#deliver_now` does:

```
# File actionmailer/lib/action_mailer/message_delivery.rb, line 117
def deliver_now
  processed_mailer.handle_exceptions do
    message.deliver
  end
end
```

File: spec/forms/schools/add_participants/add_wizard_spec.rb [hunk 1]
```diff
@@ -243,28 +238,33 @@
...
+          expect { form.save! }.not_to have_enqueued_mail(ParticipantMailer, :sit_has_added_and_validated_participant)
         end
       end
 
       context "A SIT is adding themselves as a mentor and type is set to self" do
         it "does not send the SIT an email saying they have been added and validated" do
           form.type = :self
-          form.save!
-          expect(ParticipantMailer).not_to have_received(:sit_has_added_and_validated_participant)
+          expect { form.save! }.not_to have_enqueued_mail(ParticipantMailer, :sit_has_added_and_validated_participant)
```
Comment (ltello) [inline spec/forms/schools/add_participants/add_wizard_spec.rb:267]: Lovely. Much better using :have_enqueued_mail matcher!

File: app/mailers/appropriate_body_profile_mailer.rb [hunk 2]
```diff
@@ -3,7 +3,7 @@
 class AppropriateBodyProfileMailer < ApplicationMailer
   WELCOME_TEMPLATE_ID = "0835a51f-dd7e-4a5e-b6e2-0b143de02eeb"
 
-  def welcome(appropriate_body_profile:)
+  def welcome(appropriate_body_profile)
```
Comment (ltello) [inline app/mailers/appropriate_body_profile_mailer.rb:6]: Cool. No need for keyword param here!

**Review by ltello — APPROVED** [review]
You must have had a lot of fun with the mailers, no? :-)

File: spec/forms/schools/add_participants/add_wizard_spec.rb (hunk 1 above)
Comment (peteryates) [reply spec/forms/schools/add_participants/add_wizard_spec.rb:267]: Thank @slorek for tidying all these up! 😅 👏🏽

**Pull request merged**